### ISSUE TITLE
Avoid flicker when pdf is reloaded

### ISSUE THREAD

**Issue by a-reisberg** [texstudio-org/texstudio#572]
Whenever the pdf viewer reloads the pdf file (say, after a compilation), the screen flickers, which is quite distracting. So I've been playing with the code to try to circumvent this problem. I found something interesting but I don't quite understand. Probably you could give me some pointers.

In `PDFDocument.cpp`, if I comment out the following two lines
```C++
reloadPage();
windowResized();
```
then obviously, the pdf file doesn't get rendered anymore. But I can get it to load initially by scrolling up and down (my test pdf file has multiple pages). The interesting thing is that, now, suppose I recompile the file, and then double click on the pdf viewer to briefly activate zoom, then the page gets reloaded without and flicker. The content gets updated without any side effect.

So it seems like when content is reloaded, the viewer gets white-painted, then some processing is done before actual content is painted. This white-painting doesn't happen when magnifying glass is used. But I cannot pin down precisely how this happens to make the rendering part of content reloading sequence more like that of the magnifying glass tool. Your help would be really appreciated!

Thanks!!!

**Comment by a-reisberg** [comment]
So after digging a bit more, I've narrowed down the issue a quite a bit. Essentially, within the reload sequence, whenever `update()` is called, the view flickers. I tried the following, which seemed to work most of the time (exactly when it works is written below):

- Comment out `windowResized();` inside `PDFWidget::setDocument` while keeping with `reloadPage`, just to make things simple.
- Inside `reloadPage`, comment out `update()` and replace it with the following
```C++
 if (!magnifier) magnifier = new PDFMagnifier(this, dpi);
 magnifier->show();
 magnifier->hide();
```

Then, provided that `magnifier` has been actually used before (by the user), and provided that only **one** pdf file is being displayed, the reload is completely seamless! If more than one pages are being displayed, some of the pages flicker some don't. This is quite puzzling.

Any insights would be really appreciated!

**Comment by a-reisberg** [comment]
So I finally figured out how to make it not flicker anymore. Essentially, instead of running `update()` at file reload, just add and remove a dummy `QWidget` to `PDFWidget`. This will also trigger a repaint from Qt, but somehow, Qt handles this much more gracefully. The reload now is completely smooth!

The code is here: a-reisberg/texstudio@53d46e8702d90384420b716226db4d8ee769a556

Is there a better way to do this? If not, should I make a pull-request?

Thanks!

**Comment by sunderme** [comment]
I would suggest that you find a more traditional solution. I would integrate a "hands-on" work-around only to avoid fatal flaws, not for cosmetics.

**Comment by a-reisberg** [comment]
@sunderme Of course! That's what I'm trying to figure out. I have zero experience with qt. Do you have any suggestions? Thanks!

**Comment by a-reisberg** [comment]
So here's the problem: when `update()` is called, qt really can't wait to do the repaint, so somehow, double buffering technique is not really employed (at least properly) and qt erases stuff too early. So if `paintEvent` takes too long, we will see the white paint.

To fix, I created a new function called `delayedUpdate()` and replace it in places where flicker is caused by calling `update()`. All that `delayedUpdate()` does is to fill the cache first before calling the real `update()`. The result is a completely smooth reloading experience!

The code is here: a-reisberg/texstudio@4c324b9eea3d3d63a4ea869bf9de74c4bd8f6f3b
Should I make a pull request?

**Comment by sunderme** [comment]
hmmm, the cache recalculated for all pages for every update, even for cursor movement (here goto first etc.)
For larger documents with a full cache, this will slow things down considerably...
And it inhibits the asynchronous rendering as well.
And finally, what happens if the pages have been removed from the pdf, but the cache contains that page number ?

**Comment by a-reisberg** [comment]
If I understand correctly, `pages` only contain those that are currently displaced, not all pages. I did `qDebug()` and it's indeed the case. 

`paintEvent` in fact also loops through the pages (I copied it from there). So computationally, I'm not doing anything that `paintEvent' is not doing already. I just call `renderToImage` before so it's gonna start working on the cache before `paintEvent` is called. The logic of `paintEvent` is unchanged.

Now, the only time where I replace `update()` by `delayedUpdate()` is for non-real-time scenarios, like pdf reloading. All the other stuff, like scrolling etc., is unaffected. In fact, for pdf reload, it reduces painting time, since `renderToImage` seems sometimes return cache, and then compute, and then refresh, which makes more paint calls (which sometimes also causes multiple flickers). Calling it first seems to reduce the number of repainting necessary. And as the logic of `paintEvent` stays the same, 

Finally, whether cache contains that page number doesn't affect it in any way. This will just signal renderer to recompute the cache before, rather than inside, `paintEvent()`. Later on, when `paintEvent` is called, the same logic as before applies.

**Comment by a-reisberg** [comment]
Also, github is displaying git diff in a confusing way. `PDFWidget::goFirst` is unchanged. The thing I added was for `PDFWidget::goToPageDirect` but it's displayed right below, which looks like it's part of `goFirst`. `goToPageDirect`, if I understand correctly from the name, is not time-sensitive. But if it is, removing it is fine (see the end of my previous comment).

**Comment by a-reisberg** [comment]
So the problem now is that `windowResized` is called too many times during use and `delayedUpdate` does slow things down. But `windowResized` is also used during pdf reload and causes flickers. So I'll try to change the pdf reload sequence and leave `windowResized` unchanged.

**Comment by sunderme** [comment]
okay, I understand it a bit better.
Are youmsure that pages are not rendered twice if the cache is not yet filled for repaint ?

**Comment by a-reisberg** [comment]
You surely understand it better than me. I only started looking at the code yesterday. So I haven't really read the rendering engine carefully. But it seems that by calling `renderToImage` with `priority` and `cache` options, it will be filled before `update()` is called. So no pages should get rendered twice.

However, since the calls to `renderToImage` might take a long time, it might slow down things. For example, if you have

```C++
update(); update(); update();
```
Before, qt would just call `update()` once, and all heavy work is done once in `paintEvent`. Now, some of the work is done outside of `update()`. So it effectively becomes
```C++
render(); update(); render(); update(); render(); update();
```
So if we're cached, it's the same as before, since 2nd and 3rd `render`'s should return pretty fast. However, if we miss cache (i.e. when we scroll), it'll be slow.

`windowResized` seems to always be called via `setGrid` and hence, whenever we scroll. Right?

**Comment by a-reisberg** [comment]
I've been tuning and testing the code and I've finally got something that is fast (I can't notice any difference between this version and the original tex studio) and has no flicker on file load. I've tested on Windows, Mac, and Linux, and it's all working fine.

As before, the code is here: https://github.com/texstudio-org/texstudio/compare/master...a-reisberg:NoFlicker2

Travis builds can also be found here if you want to test it quickly
https://github.com/a-reisberg/texstudio/releases

Thanks!

**Comment by a-reisberg** [comment]
Just some more comments. 

Inside `delayedUpdate()`, `PDFWidget::paintEvent` are called (exactly those calls that would happen inside `paintEvent` regardless), and then, `update()` is called after some delay. This delay is set by a variable, which is `250` on pdf reload, and `10` otherwise. `delayedUpdate()` is essentially only used in places that are related to pdf reload. And when it's happen to be called outside of the reload (like in window resized), the `10` delay does not affect performance, and in fact, sometimes it makes things seem more smoothly.

And here is the sequence of events when `delayedUpdate()` is called (as far as I understand it):

- If the rendered images are cached, `PDFWidget::paintEvent` doesn't really do much, and then `update()` would be called as usual. The delay `250` used on pdf reload is unnoticeable, and so is the otherwise `10` delay (I could set this to `0`, but `10` seems to make it work more smoothly).
- If the render images aren't in cache, they will get queued to be filled, so that the cache is better primed when `paintEvent` is called. When it's called again in `paintEvent`, `renderManager` does actually check if something gets enqueued already, so that things don't get rendered twice.

**Comment by sunderme** [comment]
I am not a fan of "global" variables.
Why don't you just add a parameter "delayDuration" to delayedUpdate with a default of 10 , which is set to higher values where needed.

**Comment by a-reisberg** [comment]
Me neither! But `delayedDuration` would then need to be passed to `paintEvent`, `windowResized`, and also the functions that `windowsResized` calls. So I would need to change the signatures of all of them (of course, with defaults, but still). That was what I did at first, but it seems to pollute the codebase more without making it any clearer. Essentially, we want to have two different behaviors depending on whether a pdf file is being reloaded. So I thought that it makes sense to have a "global" variable which records the state of the object.

But if you prefer explicit argument passing, I'm happy to fix.

**Comment by sunderme** [comment]
I doubt that it is passed to paintEvent.
paintEvent is called by the event loop after being told to do so via the command "update()".

What you want to achieve is that the rendering is already done when paintEvent is called, because paintEvent draws a blank rectangle before replacing it with actual rendered page. (that is where the flicker results from)

**Comment by a-reisberg** [comment]
Sorry I meant `reloadPage`.

**Comment by a-reisberg** [comment]
Ok, here should be the final version (but of course, I'm happy to fix if you have any suggestion for improvement).

Code is here: https://github.com/texstudio-org/texstudio/compare/master...a-reisberg:NoFlicker2

I read through the renderer and figured out how to do it more properly. Now, no more global variable, no more timer/delay (see comments in code for more details). I also improved the loading a bit comparing to original version. Essentially, when renderer calls back, at most the page that is involved is updated rather than the whole thing. This reduces the number of calls to `update()` and also makes the screen update faster in general. 

I've tested on Mac, Windows, and Ubuntu. If you want to give it a quick run, you can find it here: https://github.com/a-reisberg/texstudio/releases/tag/2.12.16beta2-NoFlicker.9.

**Comment by sunderme** [comment]
unfortunately your code does not work properly.

1. in a fully cached state, no update is called at all. This is noticable when you want to jump to the next page/previous page via the toolbar icon
2. if the rendering takes a long time (I have seen >30s for certain pages), nothing is drawn. The idea was, to place a blank page and then replace it with the correct page when it is rendered. Hence the delayed call to update (single shot timer) had some merits ...

**Comment by a-reisberg** [comment]
Thanks for the feedback!!!

I fixed it here:
https://github.com/texstudio-org/texstudio/compare/master...a-reisberg:NoFlicker2
CI builds should be ready soon here:
https://github.com/a-reisberg/texstudio/releases/tag/2.12.16beta2-NoFlicker.10

I added a delayMode flag to `renderToImage`. This fixed the concerns that you raised, and everything else seems to work like before.

Thanks again.

**Comment by sunderme** [comment]
i would prefer the timer to be something like 'QTimer::singleShot(previewCompileTimeOut, p1, SLOT(kill()));'
Not sure that 'new timer...' will generate a memory leak ...

**Comment by a-reisberg** [comment]
Sure. I just saw a lot of new timers in the code already so I thought that it's ok. So before I make another commit, this is ok?
```C++
QTimer::singleShot(previewCompileTimeOut, this, SLOT(update()));
```

**Comment by sunderme** [comment]
looks okay.
maybe last thing to consider, for every delayUpdate a timer is set and after a second update is called, whatever the current situation is.
So first, no delayed update is needed if everything is already cached.
Secondly one could check whether the per timer update is still needed when called because maybe the image is already rendered and called update itself ...

**Comment by a-reisberg** [comment]
Great! :)

It seems to me that repainting a cached thing is pretty cheap, and painting usually happens pretty often, so probably we don't need this extra complication. Probably there's no much gain (compared to the idle file loading situation, which is where this sort of technique is used here). What do you think?

**Comment by sunderme** [comment]
at least the first point is not difficult to realize.

**Comment by a-reisberg** [comment]
I actually through about the first point a bit, and it's a bit complicated in the async situation, especially when you have multiple pages involved. That was why I was pretty happy to realize that I could just paint one page each time and avoid doing that.

One way to go about this is to do the following: inside renderToImage, replace argument `bool delayMode` to `int delayTimeout` and try the following
```C++
// delayTimeout = -1 means it's NOT been called by delayedUpdate
// delayTimeout >= 0 means it's been called called by delayedUpdate and delayedUpdate wants to force an update after delayTimeout
if (delayTimeout >= 0)
    if (!img.isNull())
        QMetaObject::invokeMethod(info.obj, info.slot, Q_ARG(QPixmap, img), Q_ARG(int, pageNr));
    else {
        // invokeMethod after timeout here ...
        QTimer::singleShot(delayTimeout, info.obj, ...)
    }
```
This should address the concerns you raised. But I know neither C++ nor Qt well enough to know what to put in the `else` statement. If you tell me, I can put it in :).

**Comment by sunderme** [comment]
seems feasible

**Comment by a-reisberg** [comment]
So apparently it can be done in Qt5 but not Qt4 or below
https://stackoverflow.com/questions/9922005/using-any-c-function-as-a-qt-slot

So let me just leave it as it is for now and make a pull request?

**Comment by sunderme** [comment]
alright

**Comment by a-reisberg** [comment]
Actually I've done it (for qt5)! It's much nicer now :)

https://github.com/texstudio-org/texstudio/compare/master...a-reisberg:NoFlicker2

**Comment by sunderme** [comment]
have you checked that the singleshot works ?
lambda as slot looks wrong ...

**Comment by a-reisberg** [comment]
I copied  it from here: https://wiki.qt.io/New_Signal_Slot_Syntax
At least, next page via toolbar works even though it was broken before. And that is the only place where painting can happen when we have a cache hit.

**Comment by a-reisberg** [comment]
My comment above was nonsense. Sorry. Let me test it.

**Comment by a-reisberg** [comment]
Yup. I printed stuff in there with `qDebug` and it's there.

**Comment by a-reisberg** [comment]
Ok. Seems to work great now.  🤞 
I'm make a pull request.

**Comment by timhoffm** [comment]
Fixed by #577.
